google/cloud/spanner_v1/session.py

```diff
@@ -458,47 +463,100 @@ def run_in_transaction(self, func, *args, **kw):
         )
         attempts = 0
 
-        while True:
-            if self._transaction is None:
-                txn = self.transaction()
-                txn.transaction_tag = transaction_tag
-                txn.exclude_txn_from_change_streams = exclude_txn_from_change_streams
-            else:
-                txn = self._transaction
+        observability_options = getattr(self._database, "observability_options", None)
+        with trace_call(
+            "CloudSpanner.Session.run_in_transaction",
+            self,
+            observability_options=observability_options,
+        ) as span:
+            while True:
+                if self._transaction is None:
+                    add_span_event(span, "Creating Transaction")
+                    txn = self.transaction()
+                    txn.transaction_tag = transaction_tag
+                    txn.exclude_txn_from_change_streams = (
+                        exclude_txn_from_change_streams
+                    )
+                else:
+                    txn = self._transaction
 
-            try:
                 attempts += 1
-                return_value = func(txn, *args, **kw)
-            except Aborted as exc:
-                del self._transaction
-                _delay_until_retry(exc, deadline, attempts)
-                continue
-            except GoogleAPICallError:
-                del self._transaction
-                raise
-            except Exception:
-                txn.rollback()
-                raise
-
-            try:
-                txn.commit(
-                    return_commit_stats=self._database.log_commit_stats,
-                    request_options=commit_request_options,
-                    max_commit_delay=max_commit_delay,
-                )
-            except Aborted as exc:
-                del self._transaction
-                _delay_until_retry(exc, deadline, attempts)
-            except GoogleAPICallError:
-                del self._transaction
-                raise
-            else:
-                if self._database.log_commit_stats and txn.commit_stats:
-                    self._database.logger.info(
-                        "CommitStats: {}".format(txn.commit_stats),
-                        extra={"commit_stats": txn.commit_stats},
+                span_attributes = {"attempt": attempts}
```

Comment:
I wonder that this span attribute keeps getting overriden during retries?
Instead should we add an event saying `Starting attempt0`, `Starting attempt1` etc?

Reply:
I won't get overwritten at all: every time that this loop runs, attempt is incremented and when that you invoke span.add_event that's a new event created it isn't like replacing a value in a map.